app/code/Magento/Catalog/Model/ResourceModel/Product/Collection.php

```diff
@@ -1644,7 +1644,7 @@ public function addAttributeToSort($attribute, $dir = self::SORT_ORDER_ASC)
                 return $this;
             }
             if ($this->isEnabledFlat()) {
-                $this->getSelect()->order("cat_index_position {$dir}");
+                $this->getSelect()->order("cat_index.position {$dir}");
```

Comment:
I don't see how this fix helps. Because taking into account issue details https://github.com/magento/magento2/issues/5504
SQL query which fires Exception is 
```
Exception #0 (Zend_Db_Statement_Exception): SQLSTATE[42S22]: Column not found: 1054 Unknown column 'cat_index_position' in 'order clause', query was: 
SELECT 1 AS `status`, `e`.`entity_id`, `e`.`attribute_set_id`, `e`.`type_id`, `e`.`created_at`, `e`.`updated_at`, `e`.`sku`, `product_entity`.`row_id`, `stock_status_index`.`stock_status` AS `is_salable` 
FROM `catalog_product_flat_3` AS `e`
INNER JOIN 
	`catalog_product_entity` AS `product_entity` ON product_entity.entity_id = e.entity_id AND (product_entity.created_in <= 1 AND product_entity.updated_in > 1)
INNER JOIN 
	`cataloginventory_stock_status` AS `stock_status_index` ON e.entity_id = stock_status_index.product_id AND stock_status_index.website_id = 0 AND stock_status_index.stock_id = 1 
ORDER BY `cat_index_position` ASC, `e`.`entity_id` ASC
 LIMIT 102 exception(s):
```
and  there is no neither table not alias with a name **cat_index** in current query